compiler/rustc_middle/src/ty/mod.rs

```diff
@@ -1002,19 +1017,118 @@ impl<'tcx> ParamEnv<'tcx> {
     ///
     /// [param_env_guide]: https://rustc-dev-guide.rust-lang.org/typing_parameter_envs.html
     #[inline]
-    pub fn empty() -> Self {
-        Self::new(ListWithCachedTypeInfo::empty())
+    pub fn empty(tcx: TyCtxt<'tcx>) -> Self {
+        tcx.mk_param_env(ParamEnvInner {
+            trait_clauses: ListWithCachedTypeInfo::empty(),
+            region_outlives_clauses: ListWithCachedTypeInfo::empty(),
+            type_outlives_clauses: ListWithCachedTypeInfo::empty(),
+            projection_clauses: ListWithCachedTypeInfo::empty(),
+            const_arg_has_type_clauses: ListWithCachedTypeInfo::empty(),
+            well_formed_clauses: ListWithCachedTypeInfo::empty(),
+            const_evaluatable_clauses: ListWithCachedTypeInfo::empty(),
+            host_effect_clauses: ListWithCachedTypeInfo::empty(),
+        })
```

Comment:
Maybe cache this in `tcx`? Or would it be possible to put it in a static and insert this static in the param env interner such that you don't need to pass `tcx` around in many cases?

Reply:
I've looked at doing this locally and it has negligible performance impact as far as I can measure, so I think this would only be worth doing if we wanted to avoid passing `tcx` in.